File: LeetcodePY/array/JZ53.py
# 统计一个数字在排序数组中出现的次数。
# class Solution:
#     def search(self, nums: [int], target: int) -> int:
#         def helper(tar):
#             i, j = 0, len(nums) - 1
#             while i <= j:
#                 m = (i + j) // 2
#                 if nums[m] <= tar:
#                     i = m + 1
#                 else:
#                     j = m - 1
#             return i
#
#         return helper(target) - helper(target - 1)

# 一个长度为n-1的递增排序数组中的所有数字都是唯一的，并且每个数字都在范围0～n-1之内。在范围0～n-1内的n个数字中有且只有一个数字不在该数组中，请找出这个数字。
class Solution:
    def missingNumber(self, nums: [int]) -> int:
        i, j = 0, len(nums) - 1
        while i <= j:
            m = (i + j) // 2
            if nums[m] == m:
                i = m + 1
            else:
                j = m - 1
        return i

File: src/LeetCode/arrays/JZ53.java
package LeetCode.arrays;

public class JZ53 {
    /**
     * 统计一个数字在排序数组中出现的次数。
     */
//    class Solution {
//        public int search(int[] nums, int target) {
//            int count = 0;
//            for (int n : nums
//            ) {
//                if (n == target) {
//                    count++;
//                }
//            }
//            return count;
//        }
//    }
//    class Solution {
//        public int search(int[] nums, int target) {
//            return helper(nums, target) - helper(nums, target - 1);
//        }
//        int helper(int[] nums, int tar) {
//            int i = 0, j = nums.length - 1;
//            while(i <= j) {
//                int m = (i + j) / 2;
//                if(nums[m] <= tar) i = m + 1;
//                else j = m - 1;
//            }
//            return i;
//        }
//    }
//}

    class Solution {
        public int missingNumber(int[] nums) {
            int left = 0, right = nums.length - 1;
            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (nums[mid] == mid) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
            return left;
        }
    }
}